### PR TITLE
8/4: Add PerspectiveServiceClient and its testing

### PR DESCRIPTION
**Pull request by annemariecabs**
Add PerspectiveServiceClient, which actually calls PerspectiveAPI to analyze text. Add PerspectiveServiceClientTest, which tests PerspectiveServiceClient.

### CODE REVIEW

File: backstory/src/main/java/com/google/sps/perspective/PerspectiveManager.java [hunk 1]
```diff
@@ -39,7 +39,7 @@
       AttributeType.TOXICITY, 
       AttributeType.UNSUBSTANTIAL
     };
-
+
```
Comment (annemariecabs) [inline backstory/src/main/java/com/google/sps/perspective/PerspectiveManager.java:42]: Accidental space added. Should not affect anything about project. Please ignore.

File: backstory/src/main/java/com/google/sps/perspective/PerspectiveServiceClient.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+   *
+   * @param perspective an instance of the PerspectiveAPI to use to analyze the text for the
+   *     requested attributeTypes
+   * @param attributeTypes the requested attribute types to analyze the text for
+   * @param text the text to be analyzed by Perspective API
+   * @return an PerspectiveValues object containing all of the scores from the PerspectiveAPI
+   * @throws IllegalArgumentException if any argument is null or if text is empty
+   */
+  public static PerspectiveValues analyze(
+      PerspectiveAPI perspective, AttributeType[] attributeTypes, String text) {
```
Comment (UdayKalra) [inline backstory/src/main/java/com/google/sps/perspective/PerspectiveServiceClient.java:46]: Great edge case handling

Comment (annemariecabs) [reply backstory/src/main/java/com/google/sps/perspective/PerspectiveServiceClient.java:46]: Thanks!

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    // set up the next few calls so -1 will be returned (-1 signifies no particular value was desired).
+    when(mockResponse.getAttributeScore(any(AttributeType.class))).thenAnswer(new Answer() {
+      public Object answer(InvocationOnMock invocation) {
+        Object[] args = invocation.getArguments();
+        Object mock = invocation.getMock();
+
+        float fakeScore = -1;
+        AttributeType type = (AttributeType) args[0];
+
+        if (desiredScores.containsKey(type)) {
```
Comment (ykjchen) [inline backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java:175]: Maybe move the `float fakeScore = -1` declaration just above this line? Usually it's easier to connect pieces of code when the declaration is close to other references. This is also a strategy for reduce the scope of a variable to just where it's needed (for example, the statement `AttributeType type = (AttributeType) args[0];` doesn't need access to `fakeScore`).

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+          fakeScore = desiredScores.get(type);
+        }
+
+        when(mockAttributeScore.getSummaryScore()).thenReturn(mockScore);
+        when(mockScore.getValue()).thenReturn(fakeScore);
```
Comment (ykjchen) [inline backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java:180]: Would it be safer to create a new instance of Score and AttributeScore each time?

My concern with restubbing a single instance is that if any code holds on to an instance, it would potentially see that instance's score value change over time.

File: backstory/src/main/java/com/google/sps/perspective/PerspectiveManager.java (hunk 1 above)
Comment (annemariecabs) [reply backstory/src/main/java/com/google/sps/perspective/PerspectiveManager.java]: It didn't remove this, but I'll go do it manually.

Comment (annemariecabs) [reply backstory/src/main/java/com/google/sps/perspective/PerspectiveManager.java]: Done

File: backstory/src/main/java/com/google/sps/perspective/PerspectiveServiceClient.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+import au.com.origma.perspectiveapi.v1alpha1.models.AttributeType;
+import au.com.origma.perspectiveapi.v1alpha1.models.ContentType;
+import au.com.origma.perspectiveapi.v1alpha1.models.Entry;
+import au.com.origma.perspectiveapi.v1alpha1.models.RequestedAttribute;
+import au.com.origma.perspectiveapi.v1alpha1.models.Score;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A service client to analyze text using the Perspective API
```
[the remark this answers is not in the document]
Comment (annemariecabs) [reply backstory/src/main/java/com/google/sps/perspective/PerspectiveServiceClient.java:31]: I'll update it to ```A client``` and rename it as well. Thanks!

[the remark this answers is not in the document]
Comment (annemariecabs) [reply backstory/src/main/java/com/google/sps/perspective/PerspectiveServiceClient.java:31]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+import org.mockito.ArgumentMatchers;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+/** Quality tests for PerspectiveServiceClient */
+@RunWith(JUnit4.class)
+public final class PerspectiveServiceClientTest {
+
+  /** a PerspectiveAPI object to be used as the input api throughout the tests */
+  private static PerspectiveAPI api;
```
[the remark this answers is not in the document]
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java:45]: Will rename. Thanks!

File: backstory/src/main/java/com/google/sps/perspective/PerspectiveServiceClient.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+   * for a specified text using a specified Perspective API.
+   *
+   * @param perspective an instance of the PerspectiveAPI to use to analyze the text for the
+   *     requested attributeTypes
+   * @param attributeTypes the requested attribute types to analyze the text for
+   * @param text the text to be analyzed by Perspective API
+   * @return an PerspectiveValues object containing all of the scores from the PerspectiveAPI
+   * @throws IllegalArgumentException if any argument is null or if text is empty
+   */
+  public static PerspectiveValues analyze(
```
[the remark this answers is not in the document]
Comment (annemariecabs) [reply backstory/src/main/java/com/google/sps/perspective/PerspectiveServiceClient.java:45]: That sounds like a good idea. Will change that!

[the remark this answers is not in the document]
Comment (annemariecabs) [reply backstory/src/main/java/com/google/sps/perspective/PerspectiveServiceClient.java:45]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+   */
+  @Test
+  public void checkRightScores() {
+    // add the attributes we chose to analyze for and scores to API & create API
+
+    // add each of the desiredTypes and a random float between 0 & 1 as their score 
+    for (AttributeType type: desiredTypes) {
+      Random rand = new Random();
+      
+      desiredScores.put(type, rand.nextFloat());
```
[the remark this answers is not in the document]
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java:129]: That makes sense. I am just initializing it as a local variable now.

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    // when getAttributeScore is called we need to be able to grab the type it is looking for
+    // then, check if this AttributeType is in the map and if it is, grab its value, and
+    // set up the next few calls so that this value will be returned. if it isn't in the map,
+    // set up the next few calls so -1 will be returned (-1 signifies no particular value was desired).
+    when(mockResponse.getAttributeScore(any(AttributeType.class))).thenAnswer(new Answer() {
+      public Object answer(InvocationOnMock invocation) {
+        Object[] args = invocation.getArguments();
+        Object mock = invocation.getMock();
+
+        float fakeScore = -1;
```
[the remark this answers is not in the document]
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java:172]: That makes sense. I'll make the change.

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java (hunk 3 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java:175]: Yes, I see what you mean. I'll move it down.

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java (hunk 4 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java:180]: That makes sense. I'll just create a new instance every time.

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+   * @param desiredScores a map of the scores you want the mockAPI to return 
+   *    (if you don't set any, the mock version of this API will always return -1)
+   * @return the mocked version of the PerspectiveAPI
+   */
+  private static PerspectiveAPI createMockAPI(Map<AttributeType, Float> desiredScores) {
+    // instantiate all the necessary mock objects
+    PerspectiveAPI mockAPI = mock(PerspectiveAPI.class);
+    AnalyzeCommentResponse mockResponse = mock(AnalyzeCommentResponse.class);
+    AttributeScore mockAttributeScore = mock(AttributeScore.class);
+    Score mockScore = mock(Score.class);
```
[the remark this answers is not in the document]
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java:158]: I see what you mean. I think I'll change it so that the simpler ones (specifically AttributeScore and Score) are just constructed not mocked.

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java (hunk 4 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java (hunk 3 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveServiceClientTest.java]: Done

File: backstory/src/main/java/com/google/sps/perspective/PerspectiveServiceClient.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (annemariecabs) [reply backstory/src/main/java/com/google/sps/perspective/PerspectiveServiceClient.java]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+import org.mockito.stubbing.Answer;
+
+/** Quality tests for PerspectiveAPIClient */
+@RunWith(JUnit4.class)
+public final class PerspectiveAPIClientTest {
+
+  /** an instance of the PerspectiveAPIClient to use throughout our tests */
+  private static PerspectiveAPIClient client;
+  /** a mock PerspectiveAPI object to be used as the input api throughout the tests */
+  private static PerspectiveAPI mockAPI;
```
Comment (ykjchen) [inline backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:47]: Can these two be `private` or `private final` instead of `private static`? JUnit creates a separate instance of this class for each test case, and it would be less error prone to create a new instance for each test, rather than share an instance across multiple tests (in case state gets carried from test to test):
https://stackoverflow.com/a/43730364

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+    // set up the next few calls so that this value will be returned. if it isn't in the map,
+    // set up the next few calls so -1 will be returned (-1 signifies no particular value was desired).
+    when(mockResponse.getAttributeScore(any(AttributeType.class))).thenAnswer(new Answer() {
+      public Object answer(InvocationOnMock invocation) {
+        Object[] args = invocation.getArguments();
+        Object mock = invocation.getMock();
+
+        AttributeType type = (AttributeType) args[0];
+
+        float scoreValue = -1;
```
Comment (ykjchen) [inline backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:175]: How about setting this to a private constant, so that it seems less like a "magic number" (reader is a bit more aware what the significance of this value is)?

```
private static final DEFAULT_ATTRIBUTE_SCORE = -1;
```

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+        if (desiredScores.containsKey(type)) {
+          scoreValue = desiredScores.get(type);
+        }
+
+        // create a mocked score with the desired scoreValue and null ScoreType
+        Score mockScore = new Score(scoreValue, null);
```
Comment (ykjchen) [inline backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:181]: `mockScore` -> `score` (since it's a real and not mock object)?

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+        // create a mocked score with mockScore as summary score and null list of spanscores
+        AttributeScore mockAttributeScore = new AttributeScore(mockScore, null);
```
Comment (ykjchen) [inline backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:183]: `mockAttributeScore` -> `attributeScore` (since it's a real and not mock object)?

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+/** Quality tests for PerspectiveAPIClient */
+@RunWith(JUnit4.class)
+public final class PerspectiveAPIClientTest {
+
+  /** an instance of the PerspectiveAPIClient to use throughout our tests */
+  private static PerspectiveAPIClient client;
+  /** a mock PerspectiveAPI object to be used as the input api throughout the tests */
+  private static PerspectiveAPI mockAPI;
+  /** the array of attribute types that we want scores back for */
+  private static final AttributeType[] desiredTypes = { AttributeType.PROFANITY, AttributeType.TOXICITY, AttributeType.UNSUBSTANTIAL };;
```
Comment (ykjchen) [inline backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:49]: `Constant names use CONSTANT_CASE`:
https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+    Assert.assertEquals(desiredTypes.length, values.getAttributeTypesToScores().size());
+  }
+
+  /**
+   * Check that analyze() returns a PerspectiveValues
+   * object with the same text as passed in.
+   */
+  @Test
+  public void checkTextSame() {
+    PerspectiveValues values = client.analyze(desiredTypes, "foo");
```
Comment (ykjchen) [inline backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:113]: Creating a local variable would make this input more obviously connected to the assertion:

```
String text = "foo";
PerspectiveValues values = client.analyze(desiredTypes, text);
Assert.assertEquals(text, values.getText());
```

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+  /** a mock PerspectiveAPI object to be used as the input api throughout the tests */
+  private static PerspectiveAPI mockAPI;
+  /** the array of attribute types that we want scores back for */
+  private static final AttributeType[] desiredTypes = { AttributeType.PROFANITY, AttributeType.TOXICITY, AttributeType.UNSUBSTANTIAL };;
+
+
+  @Before 
+  public void setUp() {
+    // will set up an API that only returns -1
+    mockAPI = createMockAPI(new HashMap<AttributeType, Float>());
```
Comment (ykjchen) [inline backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:55]: It doesn't look like this needs to be an instance variable, but can be a local variable instead.

There's only one tests that creates a different PerspectiveAPI mock, and that can be a local variable too.

(This would reduce state within the test case, which reduces room of introducing errors).

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+   * object with the right scores.
+   */
+  @Test
+  public void checkRightScores() {
+    // a Map to be used as the desired scores for the mock API to return 
+    Map desiredScores = new HashMap<AttributeType, Float>();
+
+    // add each of the desiredTypes and a random float between 0 & 1 as their score 
+    for (AttributeType type: desiredTypes) {
+      Random rand = new Random();
```
Comment (ykjchen) [inline backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:129]: `rand` -> `random`

```
do not abbreviate by deleting letters within a word
```
https://google.github.io/styleguide/jsguide.html#naming-rules-common-to-all-identifiers

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+import org.mockito.ArgumentMatchers;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+/** Quality tests for PerspectiveAPIClient */
+@RunWith(JUnit4.class)
+public final class PerspectiveAPIClientTest {
```
Comment (ykjchen) [inline backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:42]: Can you leave a TODO to add a test that checks that the correct AnalyzeCommentRequest is passed to PerspectiveAPI, and follow up with a separate PR with that test case?

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java (hunk 11 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:47]: I agree. I will make them just private.

Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:47]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java (hunk 15 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:49]: Thanks for the reminder. This was originally not a constant, but now that it is, I will change it.

Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:49]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java (hunk 13 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:181]: That makes sense. Will change this. I kept it because it technically is a part of the mock, but it is real because it's constructed with the public constructor for the class.

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java (hunk 14 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:183]: Will change this. Kept for [same reason](https://github.com/googleinterns/step174-2020/pull/132#discussion_r466814517) as mockScore was kept.

Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:183]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java (hunk 13 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:181]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java (hunk 12 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:175]: That's a great idea. Will definitely make code more comprehensible.

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java (hunk 16 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:113]: I agree. I think it might be even clearer just to have a default text constant throughout file. I think I'll add this:
`/** a piece of default text to use in tests */
  private static final String DEFAULT_TEXT = "foo";`

Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:113]: Done

**Comment by annemariecabs** [comment]
Note: I did end up changing one of the parameters of analyze() to a List rather than an array because I felt it allowed for more flexibility while maintaining the only requirement (a List with AttributeType objects).

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java (hunk 18 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:129]: Thanks for pointing that out!

Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:129]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java (hunk 12 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:175]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java (hunk 19 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:42]: Great idea! Will do.

Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:42]: Done

File: backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java (hunk 17 above)
Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:55]: That makes sense. I'll make the change.

Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:55]: Also, I think I'm going to just make client local as well as it takes mockAPI as a parameter.

Comment (annemariecabs) [reply backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java:55]: Done
